File: ARITH/arith.pl
#!/usr/bin/perl
#
use bignum;

$cases = <STDIN>;

sub print_res( )
{
	my $num1 = shift;
	my $num2 = shift;
	my $res = shift;
	my $opr = shift;

	$len = length($num1);
	if ( $len < length($num2) + 1  )
	{
		$len = length($num2) + 1 ;
	}
	if ( $len < length($res) )
	{
		$len = length($res);
	}

	#print first number 
	for ( $i = 0 ; $i<$len - length($num1) ; $i++)
	{
		print " ";
	}
	print "$num1\n" ;
	for ( $i = 0 ; $i<$len - ( length($num2) +1 ); $i++)
	{
		print " ";
	}
	print "$opr$num2\n";
	for( $i = 0 ; $i<$len ; $i++)
	{
		print "-";
	}
	print "\n";
	for ( $i = 0 ; $i<$len - ( length($res) ); $i++)
	{
		print " ";
	}
	print "$res\n";
}

sub print_with_spaces( )
{
	$lenx = shift;
	$str = shift;
	$len2 = length($str);
	#print "str : $str \n";
	for ( $i =0 ;$i< $lenx-$len2 ; $i++ )
	{
		print " ";
	}
	print $str,"\n";
}

while ($cases-- )
{
	chomp($line = <STDIN>);
	$line  =~  /(\d+)([\+\-\*])(\d+)$/;
	$num1 = $1;
	$num2 = $3;
	$opr  = $2;
	chomp($num2);

	if ( $opr eq "+" )
	{
		$res = $num1 + $num2 ;
		$opr = "+";
		&print_res($num1,$num2,$res,$opr);

	}
	elsif ( $opr eq "-" )
	{
		$res = $num1 - $num2;
		$opr = "-";
		&print_res($num1,$num2,$res,$opr);
	}
	elsif ( $opr eq "*")
	{
		$res = $num1*$num2;
		$len = length($res);
		if ( $len < length($num1) )
		{
			$len = length($num1) ;
		}
		if ( $len < length($num2) +1 )
		{
			$len = length($num2) + 1 ;
		}
		&print_with_spaces( $len , "$num1" );
		&print_with_spaces( $len , "*$num2" );
		@vals = split   // , $num2 ;
		$vallen = @vals;
		if ( $vallen > 1 ) 
		{
			$len2 = length($num1);
		 	if ( $len2 < length($num2) +1 )
			{
				$len2 = length($num2) +1 ;
			}
			$dashes = "";
			for ( $i = 0 ; $i< $len2 ; $i++ )
			{
				$dashes .="-";
			}
			&print_with_spaces( $len , $dashes );
			$j=0 ;
			foreach $val (reverse @vals)
			{
				$tmp = $val*$num1;
				#print "tmp:$tmp";
				&print_with_spaces($len - $j , $tmp);
				$j++;
				#	print "len $len \n";
			}
		}
		for ( $i = 0 ; $i< $len ; $i++ )
		{
			print "-";
		}
		print "\n";	
		&print_with_spaces($len , $res );
	}
	print "\n";

}

File: ARITH/arith.py
#!/usr/bin/python 

#stupid perl not taking long ints
# I love python  :)
import re

def print_with_space( val , nos ) :
	len2 =  len(val) 
	final = " "*(nos-len2)
	final= final+val
	print final 


regex = re.compile(r"(\d+)([\+\-\*])(\d+)$")
cases = int (raw_input())
while cases :
	line = raw_input()
	match = regex.match(line)
	num1 = match.group(1)
	num2 = match.group(3)
	opr  = match.group(2)

#	print "num1 :"  , num1
#	print "num2 :"  , num2
#	print "opr :"  , opr 

	if opr == "+" :
		res = str(int(num1) + int(num2))

		len1 = len(res)
		if len1  < len(num2) + 1  :
			len1 = len(num2) + 1

		print_with_space( num1 , len1 )
		print_with_space( "+" + num2 , len1 )
		print_with_space( "-"*len1 , len1 )
		print_with_space(res , len1 )
	if opr == "-" :
		res = str(int(num1) - int(num2))

		len1 = len(res)
		if len1  < len(num1)   :
			len1 = len(num1) 
		if len1 < len(num2) + 1 :
			len1 = len(num2) + 1 

		len2 = len(res)
		if len2 < len(num2) + 1 :
			len2 = len(num2)+1 
		
		print_with_space( num1 , len1 )
		print_with_space( "-" + num2 , len1 )
		print_with_space( "-"*len2 , len1 )
		print_with_space(res , len1 )
	if opr == "*" :

		
		res = str(int(num1) * int(num2))
		len1 = len(res)
		if len1  < len(num1)   :
			len1 = len(num1) 
		if len1 < len(num2) + 1 :
			len1 = len(num2) + 1 

		print_with_space(num1,len1 )
		print_with_space("*"+num2 , len1 )
		if ( len(num2) == 1 ) :
			len2 = max(len(num2)+1 , len(res) )
			print_with_space("-"*len2 , len1 )
			print_with_space(res , len1 )
		else :
			x = len1
			num1_int = int(num1)
			dlen = max ( len(num2)+1 , len( str( num1_int * int(num2[0]) ) ) )
			print_with_space("-"*dlen , len1 )
			for val in num2[::-1] :
				tmp = str( int(val) * num1_int)
				print_with_space(tmp , x ) 
				x-=1

			#dlen = max( len(res) , len(num2) + 1 )	
			#if  num1_int == 0 :
			#	print_with_space( "-"*len(num2) , dlen ) 
			#else :
			print_with_space( "-"*len(res) , dlen ) 
			print_with_space( res , len1 ) 

	print "\n",
	cases-=1


